File: project_runtime_settings.py
from __future__ import annotations

from typing import Any


DEFAULT_FORMAL_SAVE_SLOT_COUNT = 24
MIN_FORMAL_SAVE_SLOT_COUNT = 3
MAX_FORMAL_SAVE_SLOT_COUNT = 120
PROJECT_RUNTIME_TEXT_SPEED_CPS = {
    "slow": 24,
    "normal": 42,
    "fast": 72,
    "instant": 10000,
}
PROJECT_RUNTIME_TEXT_SPEEDS = frozenset(PROJECT_RUNTIME_TEXT_SPEED_CPS)
PROJECT_RUNTIME_DIALOG_THEMES = frozenset({"project", "warm", "moonlight", "paper", "transparent"})
PROJECT_RUNTIME_UI_THEME_MODES = frozenset({"auto", "light", "dark"})


def _clean_key(value: Any, fallback: str = "") -> str:
    text = str(value or "").strip().lower()
    return text or fallback


def clamp_int(value: Any, fallback: int, minimum: int, maximum: int) -> int:
    try:
        numeric = int(value)
    except (TypeError, ValueError):
        numeric = fallback
    return max(minimum, min(maximum, numeric))


def build_default_project_runtime_settings() -> dict:
    return {
        "formalSaveSlotCount": DEFAULT_FORMAL_SAVE_SLOT_COUNT,
        "defaultTextSpeed": "normal",
        "defaultDialogTheme": "project",
        "defaultUiThemeMode": "auto",
        "defaultBgmVolume": 72,
        "defaultSfxVolume": 85,
        "defaultVoiceVolume": 92,
        "defaultVoiceEnabled": True,
        "defaultVoiceDuckingEnabled": True,
    }


def sanitize_project_runtime_settings(value: Any) -> dict:
    source = value if isinstance(value, dict) else {}
    defaults = build_default_project_runtime_settings()
    text_speed = _clean_key(source.get("defaultTextSpeed"), defaults["defaultTextSpeed"])
    dialog_theme = _clean_key(source.get("defaultDialogTheme"), defaults["defaultDialogTheme"])
    ui_theme_mode = _clean_key(source.get("defaultUiThemeMode"), defaults["defaultUiThemeMode"])
    return {
        "formalSaveSlotCount": clamp_int(
            source.get("formalSaveSlotCount"),
            DEFAULT_FORMAL_SAVE_SLOT_COUNT,
            MIN_FORMAL_SAVE_SLOT_COUNT,
            MAX_FORMAL_SAVE_SLOT_COUNT,
        ),
        "defaultTextSpeed": text_speed if text_speed in PROJECT_RUNTIME_TEXT_SPEEDS else defaults["defaultTextSpeed"],
        "defaultDialogTheme": dialog_theme if dialog_theme in PROJECT_RUNTIME_DIALOG_THEMES else defaults["defaultDialogTheme"],
        "defaultUiThemeMode": ui_theme_mode if ui_theme_mode in PROJECT_RUNTIME_UI_THEME_MODES else defaults["defaultUiThemeMode"],
        "defaultBgmVolume": clamp_int(source.get("defaultBgmVolume"), defaults["defaultBgmVolume"], 0, 100),
        "defaultSfxVolume": clamp_int(source.get("defaultSfxVolume"), defaults["defaultSfxVolume"], 0, 100),
        "defaultVoiceVolume": clamp_int(source.get("defaultVoiceVolume"), defaults["defaultVoiceVolume"], 0, 100),
        "defaultVoiceEnabled": source.get("defaultVoiceEnabled") is not False,
        "defaultVoiceDuckingEnabled": source.get("defaultVoiceDuckingEnabled") is not False,
    }


def get_runtime_settings_from_bundle(bundle: dict | None) -> dict:
    project = bundle.get("project") if isinstance(bundle, dict) and isinstance(bundle.get("project"), dict) else {}
    return sanitize_project_runtime_settings(project.get("runtimeSettings"))


def get_effective_text_speed(block: dict | None, runtime_settings: dict | None = None) -> str:
    source = block if isinstance(block, dict) else {}
    explicit = _clean_key(source.get("textSpeed"))
    if explicit in PROJECT_RUNTIME_TEXT_SPEED_CPS:
        return explicit
    settings = sanitize_project_runtime_settings(runtime_settings)
    default_speed = settings["defaultTextSpeed"]
    # Normal is Ren'Py's readable default, so omit a generated cps tag unless the user chose another project default.
    return default_speed if default_speed != "normal" else ""


def get_effective_text_cps(block: dict | None, runtime_settings: dict | None = None) -> int | None:
    speed = get_effective_text_speed(block, runtime_settings)
    return PROJECT_RUNTIME_TEXT_SPEED_CPS.get(speed)


def get_runtime_volume_ratio(runtime_settings: dict | None, key: str) -> float:
    settings = sanitize_project_runtime_settings(runtime_settings)
    return round(settings.get(key, 100) / 100, 2)


def get_renpy_runtime_summary(runtime_settings: dict | None) -> dict:
    settings = sanitize_project_runtime_settings(runtime_settings)
    default_speed = settings["defaultTextSpeed"]
    return {
        "defaultTextSpeed": default_speed,
        "defaultTextCps": PROJECT_RUNTIME_TEXT_SPEED_CPS[default_speed],
        "defaultBgmVolume": get_runtime_volume_ratio(settings, "defaultBgmVolume"),
        "defaultSfxVolume": get_runtime_volume_ratio(settings, "defaultSfxVolume"),
        "defaultVoiceVolume": get_runtime_volume_ratio(settings, "defaultVoiceVolume"),
        "defaultVoiceEnabled": bool(settings["defaultVoiceEnabled"]),
        "defaultVoiceDuckingEnabled": bool(settings["defaultVoiceDuckingEnabled"]),
        "formalSaveSlotCount": settings["formalSaveSlotCount"],
    }

File: prototype_editor/modules/project_runtime_settings.js
(function attachProjectRuntimeSettingsTools(global) {
  "use strict";

  const DEFAULT_SAVE_SLOT_COUNT_LIMITS = Object.freeze({ min: 3, max: 120 });
  const DEFAULT_RUNTIME_SETTINGS = Object.freeze({
    formalSaveSlotCount: 24,
    defaultTextSpeed: "normal",
    defaultDialogTheme: "project",
    defaultUiThemeMode: "auto",
    defaultBgmVolume: 72,
    defaultSfxVolume: 85,
    defaultVoiceVolume: 92,
    defaultVoiceEnabled: true,
    defaultVoiceDuckingEnabled: true,
  });
  const TEXT_SPEED_CPS = Object.freeze({
    slow: 24,
    normal: 42,
    fast: 72,
    instant: 10000,
  });
  const RUNTIME_DIALOG_THEMES = Object.freeze(["project", "warm", "moonlight", "paper", "transparent"]);
  const RUNTIME_UI_THEME_MODES = Object.freeze(["auto", "light", "dark"]);

  function hasOwn(source, key) {
    return Object.prototype.hasOwnProperty.call(source ?? {}, key);
  }

  function cleanKey(value, fallback = "") {
    const text = String(value ?? "").trim().toLowerCase();
    return text || fallback;
  }

  function clamp(value, min, max) {
    return Math.min(Math.max(value, min), max);
  }

  function getSafeNumber(value, fallback = 0) {
    const number = Number(value);
    return Number.isFinite(number) ? number : fallback;
  }

  function getSafeKey(allowedKeys, value, fallback) {
    const key = cleanKey(value, fallback);
    if (Array.isArray(allowedKeys)) {
      return allowedKeys.includes(key) ? key : fallback;
    }
    return hasOwn(allowedKeys, key) ? key : fallback;
  }

  function getRuntimeDefaults(options = {}) {
    return { ...DEFAULT_RUNTIME_SETTINGS, ...(options.defaultRuntimeSettings ?? {}) };
  }

  function getSaveSlotLimits(options = {}) {
    return { ...DEFAULT_SAVE_SLOT_COUNT_LIMITS, ...(options.saveSlotCountLimits ?? {}) };
  }

  function getSafeProjectFormalSaveSlotCount(value, options = {}) {
    const limits = getSaveSlotLimits(options);
    const defaults = getRuntimeDefaults(options);
    return clamp(Math.round(getSafeNumber(value, defaults.formalSaveSlotCount)), limits.min, limits.max);
  }

  function getSafeProjectRuntimeTextSpeed(value, options = {}) {
    const defaults = getRuntimeDefaults(options);
    return getSafeKey(options.runtimeTextSpeedLabels || TEXT_SPEED_CPS, value, defaults.defaultTextSpeed);
  }

  function getSafeProjectRuntimeDialogTheme(value, options = {}) {
    const defaults = getRuntimeDefaults(options);
    return getSafeKey(options.runtimeDialogThemeLabels || RUNTIME_DIALOG_THEMES, value, defaults.defaultDialogTheme);
  }

  function getSafeProjectRuntimeUiThemeMode(value, options = {}) {
    const defaults = getRuntimeDefaults(options);
    return getSafeKey(options.runtimeUiThemeModeLabels || RUNTIME_UI_THEME_MODES, value, defaults.defaultUiThemeMode);
  }

  function getSafeProjectRuntimeVolume(value, fallback = 100) {
    return clamp(Math.round(getSafeNumber(value, fallback)), 0, 100);
  }

  function getProjectRuntimeSettings(project = {}, options = {}) {
    const runtimeSettings = project?.runtimeSettings ?? {};
    const defaults = getRuntimeDefaults(options);
    return {
      formalSaveSlotCount: getSafeProjectFormalSaveSlotCount(runtimeSettings.formalSaveSlotCount, options),
      defaultTextSpeed: getSafeProjectRuntimeTextSpeed(runtimeSettings.defaultTextSpeed, options),
      defaultDialogTheme: getSafeProjectRuntimeDialogTheme(runtimeSettings.defaultDialogTheme, options),
      defaultUiThemeMode: getSafeProjectRuntimeUiThemeMode(runtimeSettings.defaultUiThemeMode, options),
      defaultBgmVolume: getSafeProjectRuntimeVolume(runtimeSettings.defaultBgmVolume, defaults.defaultBgmVolume),
      defaultSfxVolume: getSafeProjectRuntimeVolume(runtimeSettings.defaultSfxVolume, defaults.defaultSfxVolume),
      defaultVoiceVolume: getSafeProjectRuntimeVolume(runtimeSettings.defaultVoiceVolume, defaults.defaultVoiceVolume),
      defaultVoiceEnabled: runtimeSettings.defaultVoiceEnabled !== false,
      defaultVoiceDuckingEnabled: runtimeSettings.defaultVoiceDuckingEnabled !== false,
    };
  }

  function getProjectFormalSaveSlotCount(project = {}, options = {}) {
    return getProjectRuntimeSettings(project, options).formalSaveSlotCount;
  }

  function getEffectiveTextSpeed(block = {}, runtimeSettings = {}, options = {}) {
    const explicitSpeed = getSafeProjectRuntimeTextSpeed(block?.textSpeed, {
      ...options,
      defaultRuntimeSettings: { ...getRuntimeDefaults(options), defaultTextSpeed: "" },
    });
    if (explicitSpeed) {
      return explicitSpeed;
    }
    const defaultSpeed = getProjectRuntimeSettings({ runtimeSettings }, options).defaultTextSpeed;
    return defaultSpeed === "normal" ? "" : defaultSpeed;
  }

  function getEffectiveTextCps(block = {}, runtimeSettings = {}, options = {}) {
    const speed = getEffectiveTextSpeed(block, runtimeSettings, options);
    return speed ? TEXT_SPEED_CPS[speed] : null;
  }

  function getRuntimeVolumeRatio(runtimeSettings = {}, key, options = {}) {
    const settings = getProjectRuntimeSettings({ runtimeSettings }, options);
    return Number(((settings[key] ?? 100) / 100).toFixed(2));
  }

  function getRenpyPreferenceTextCps(defaultSpeed) {
    return defaultSpeed === "instant" ? 0 : TEXT_SPEED_CPS[defaultSpeed];
  }

  function getRenpyRuntimeSummary(runtimeSettings = {}, options = {}) {
    const settings = getProjectRuntimeSettings({ runtimeSettings }, options);
    const voiceVolume = settings.defaultVoiceEnabled ? getRuntimeVolumeRatio(settings, "defaultVoiceVolume", options) : 0;
    return {
      defaultTextSpeed: settings.defaultTextSpeed,
      defaultTextCps: TEXT_SPEED_CPS[settings.defaultTextSpeed],
      renpyPreferenceTextCps: getRenpyPreferenceTextCps(settings.defaultTextSpeed),
      defaultBgmVolume: getRuntimeVolumeRatio(settings, "defaultBgmVolume", options),
      defaultSfxVolume: getRuntimeVolumeRatio(settings, "defaultSfxVolume", options),
      defaultVoiceVolume: getRuntimeVolumeRatio(settings, "defaultVoiceVolume", options),
      effectiveVoiceVolume: voiceVolume,
      defaultVoiceEnabled: settings.defaultVoiceEnabled,
      defaultVoiceDuckingEnabled: settings.defaultVoiceDuckingEnabled,
      formalSaveSlotCount: settings.formalSaveSlotCount,
    };
  }

  global.CanvasiaEditorProjectRuntimeSettings = Object.freeze({
    DEFAULT_SAVE_SLOT_COUNT_LIMITS,
    DEFAULT_RUNTIME_SETTINGS,
    TEXT_SPEED_CPS,
    RUNTIME_DIALOG_THEMES,
    RUNTIME_UI_THEME_MODES,
    getSafeProjectFormalSaveSlotCount,
    getSafeProjectRuntimeTextSpeed,
    getSafeProjectRuntimeDialogTheme,
    getSafeProjectRuntimeUiThemeMode,
    getSafeProjectRuntimeVolume,
    getProjectRuntimeSettings,
    getProjectFormalSaveSlotCount,
    getEffectiveTextSpeed,
    getEffectiveTextCps,
    getRuntimeVolumeRatio,
    getRenpyPreferenceTextCps,
    getRenpyRuntimeSummary,
  });
})(typeof window !== "undefined" ? window : globalThis);
